File: backend/routes/medicalProfile.js
const express = require('express');
const router = express.Router();
const supabase = require('../supabaseClient');

// ─────────────────────────────────────────────────────────────
// HELPER: extract userId from Clerk JWT or custom header
// ─────────────────────────────────────────────────────────────
function getUserId(req) {
  // Try Clerk auth (set by clerkAuth middleware)
  if (req.auth?.userId) return req.auth.userId;
  // Fallback: custom header for direct API calls
  if (req.headers['x-user-id']) return req.headers['x-user-id'];
  return null;
}

// ─────────────────────────────────────────────────────────────
// GET /api/medical-profile — fetch the logged-in user's profile
// ─────────────────────────────────────────────────────────────
router.get('/', async (req, res) => {
  try {
    const userId = getUserId(req);
    if (!userId) return res.status(401).json({ success: false, message: 'Unauthorized' });

    const { data, error } = await supabase
      .from('user_medical_profiles')
      .select('*')
      .eq('user_id', userId)
      .single();

    if (error && error.code === 'PGRST116') {
      // No profile yet — return empty object so frontend can show the form
      return res.json({ success: true, profile: null, hasProfile: false });
    }
    if (error) throw error;

    return res.json({ success: true, profile: data, hasProfile: true });
  } catch (err) {
    console.error('medical-profile GET error:', err.message);
    res.status(500).json({ success: false, message: err.message });
  }
});

// ─────────────────────────────────────────────────────────────
// POST /api/medical-profile — create or update profile (upsert)
// ─────────────────────────────────────────────────────────────
router.post('/', async (req, res) => {
  try {
    const userId = getUserId(req);
    if (!userId) return res.status(401).json({ success: false, message: 'Unauthorized' });

    const {
      full_name, age, date_of_birth, gender, blood_group, height_cm, weight_kg,
      conditions, bp_systolic, bp_diastolic, sugar_level_fasting, sugar_level_pp,
      pulse_rate, current_medications, allergies, preferred_hospitals,
      emergency_contact_name, emergency_contact_phone, emergency_contact_relation,
      has_insurance, insurance_provider, policy_number, organ_donor, doctor_notes,
    } = req.body;

    const profileData = {
      user_id: userId,
      full_name, age, date_of_birth: date_of_birth || null,
      gender, blood_group, height_cm: height_cm || null, weight_kg: weight_kg || null,
      conditions: conditions || [],
      bp_systolic: bp_systolic || null, bp_diastolic: bp_diastolic || null,
      sugar_level_fasting: sugar_level_fasting || null,
      sugar_level_pp: sugar_level_pp || null,
      pulse_rate: pulse_rate || null,
      current_medications: current_medications || [],
      allergies: allergies || [],
      preferred_hospitals: preferred_hospitals || [],
      emergency_contact_name, emergency_contact_phone, emergency_contact_relation,
      has_insurance: has_insurance || false,
      insurance_provider: insurance_provider || null,
      policy_number: policy_number || null,
      organ_donor: organ_donor || false,
      doctor_notes: doctor_notes || null,
    };

    const { data, error } = await supabase
      .from('user_medical_profiles')
      .upsert(profileData, { onConflict: 'user_id' })
      .select()
      .single();

    if (error) throw error;
    return res.json({ success: true, profile: data });
  } catch (err) {
    console.error('medical-profile POST error:', err.message);
    res.status(500).json({ success: false, message: err.message });
  }
});

// ─────────────────────────────────────────────────────────────
// GET /api/medical-profile/emergency-card — compact card for SOS
// Returns only the data needed for an emergency handoff
// ─────────────────────────────────────────────────────────────
router.get('/emergency-card', async (req, res) => {
  try {
    const userId = getUserId(req);
    if (!userId) return res.status(401).json({ success: false, message: 'Unauthorized' });

    const { data, error } = await supabase
      .from('user_medical_profiles')
      .select(`
        full_name, age, gender, blood_group,
        conditions, allergies, current_medications,
        bp_systolic, bp_diastolic, sugar_level_fasting,
        pulse_rate, preferred_hospitals,
        emergency_contact_name, emergency_contact_phone,
        has_insurance, insurance_provider, organ_donor,
        doctor_notes
      `)
      .eq('user_id', userId)
      .single();

    if (error && error.code === 'PGRST116') {
      return res.json({ success: true, card: null });
    }
    if (error) throw error;

    // Format as a clean emergency card
    const card = {
      ...data,
      bp: data.bp_systolic && data.bp_diastolic ? `${data.bp_systolic}/${data.bp_diastolic} mmHg` : null,
      sugar: data.sugar_level_fasting ? `${data.sugar_level_fasting} mg/dL (fasting)` : null,
      conditions_summary: (data.conditions || []).join(', '),
      medications_summary: (data.current_medications || []).join(', '),
      allergies_summary: (data.allergies || []).join(', '),
      preferred_hospital_primary: (data.preferred_hospitals || [])[0] || 'Apollo Emergency Care',
    };

    return res.json({ success: true, card });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
});

module.exports = router;

File: backend/routes/medical_profile.py
"""
routes/medical_profile.py — Patient medical profile CRUD and SOS emergency card
Mirrors backend/routes/medicalProfile.js
"""
from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel
from typing import Optional, List, Any

from middleware.auth import get_current_user
from core.supabase_client import supabase

router = APIRouter(prefix="/api/medical-profile", tags=["Medical Profile"])
TABLE = "user_medical_profiles"


class MedicalProfileRequest(BaseModel):
    full_name: Optional[str] = None
    age: Optional[int] = None
    date_of_birth: Optional[str] = None
    gender: Optional[str] = None
    blood_group: Optional[str] = None
    height_cm: Optional[float] = None
    weight_kg: Optional[float] = None
    conditions: Optional[List[str]] = []
    bp_systolic: Optional[int] = None
    bp_diastolic: Optional[int] = None
    sugar_level_fasting: Optional[float] = None
    sugar_level_pp: Optional[float] = None
    pulse_rate: Optional[int] = None
    current_medications: Optional[List[str]] = []
    allergies: Optional[List[str]] = []
    preferred_hospitals: Optional[List[str]] = []
    emergency_contact_name: Optional[str] = None
    emergency_contact_phone: Optional[str] = None
    emergency_contact_relation: Optional[str] = None
    has_insurance: Optional[bool] = False
    insurance_provider: Optional[str] = None
    policy_number: Optional[str] = None
    organ_donor: Optional[bool] = False
    doctor_notes: Optional[str] = None


@router.get("")
@router.get("/")
async def get_medical_profile(user: dict = Depends(get_current_user)):
    user_id = user["id"]
    try:
        res = supabase.from_(TABLE).select("*").eq("user_id", user_id).execute()
        if not res.data:
            return {"success": True, "profile": None, "hasProfile": False}
        return {"success": True, "profile": res.data[0], "hasProfile": True}
    except Exception as e:
        if "does not exist" in str(e):
            return {"success": True, "profile": None, "hasProfile": False}
        raise HTTPException(status_code=500, detail=str(e))


@router.post("")
@router.post("/")
async def save_medical_profile(payload: MedicalProfileRequest, user: dict = Depends(get_current_user)):
    user_id = user["id"]
    try:
        profile_data = payload.model_dump()
        profile_data["user_id"] = user_id

        res = supabase.from_(TABLE).upsert(profile_data, on_conflict="user_id").select().single().execute()
        return {"success": True, "profile": res.data}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/emergency-card")
async def get_emergency_card(user: dict = Depends(get_current_user)):
    user_id = user["id"]
    try:
        res = supabase.from_(TABLE).select(
            "full_name, age, gender, blood_group, conditions, allergies, current_medications, "
            "bp_systolic, bp_diastolic, sugar_level_fasting, pulse_rate, preferred_hospitals, "
            "emergency_contact_name, emergency_contact_phone, has_insurance, insurance_provider, "
            "organ_donor, doctor_notes"
        ).eq("user_id", user_id).execute()

        if not res.data:
            return {"success": True, "card": None}

        data = res.data[0]
        bp = f"{data['bp_systolic']}/{data['bp_diastolic']} mmHg" if data.get("bp_systolic") and data.get("bp_diastolic") else None
        sugar = f"{data['sugar_level_fasting']} mg/dL (fasting)" if data.get("sugar_level_fasting") else None

        card = {
            **data,
            "bp": bp,
            "sugar": sugar,
            "conditions_summary": ", ".join(data.get("conditions") or []),
            "medications_summary": ", ".join(data.get("current_medications") or []),
            "allergies_summary": ", ".join(data.get("allergies") or []),
            "preferred_hospital_primary": (data.get("preferred_hospitals") or ["Apollo Emergency Care"])[0],
        }

        return {"success": True, "card": card}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
